File: src/target_edit_page.py
# target_edit_page.py
#
# Copyright 2026 Unknown
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
#
# SPDX-License-Identifier: GPL-3.0-or-later

from gi.repository import Adw, Gtk, GLib, Gio
from .db import get_db


@Gtk.Template(resource_path='/io/github/nico359/cloudsend/target_edit_page.ui')
class CloudsendTargetEditPage(Adw.NavigationPage):
    __gtype_name__ = 'CloudsendTargetEditPage'

    local_folder_row = Gtk.Template.Child()
    remote_folder_row = Gtk.Template.Child()
    mirror_switch = Gtk.Template.Child()
    save_button = Gtk.Template.Child()
    delete_button = Gtk.Template.Child()

    def __init__(self, target=None, targets_page=None, **kwargs):
        """
        target: dict from db (None = creating new target)
        targets_page: reference to the targets list page (to refresh after save)
        """
        super().__init__(**kwargs)
        self._target = target
        self._targets_page = targets_page
        self._local_path = None
        self._remote_path = None

        self.local_folder_row.connect('activated', self._on_local_folder_clicked)
        self.remote_folder_row.connect('activated', self._on_remote_folder_clicked)
        self.save_button.connect('clicked', self._on_save_clicked)
        self.delete_button.connect('clicked', self._on_delete_clicked)

        if target:
            self.set_title("Edit Target")
            self._local_path = target['local_path']
            self._remote_path = target['remote_path']
            self.local_folder_row.set_subtitle(target['local_path'])
            self.remote_folder_row.set_subtitle(target['remote_path'])
            self.mirror_switch.set_active(target['mode'] == 'mirror')
            self.delete_button.set_visible(True)

    def _on_local_folder_clicked(self, row):
        dialog = Gtk.FileDialog()
        dialog.set_title("Select Local Folder")

        dialog.select_folder(
            self.get_ancestor(Gtk.Window),
            None,
            self._on_local_folder_selected,
        )

    def _on_local_folder_selected(self, dialog, result):
        try:
            folder = dialog.select_folder_finish(result)
            if folder:
                self._local_path = folder.get_path()
                self.local_folder_row.set_subtitle(self._local_path)
        except GLib.Error:
            pass  # User cancelled

    def _on_remote_folder_clicked(self, row):
        from .remote_browser_page import CloudsendRemoteBrowserPage

        initial_path = self._remote_path or '/'
        page = CloudsendRemoteBrowserPage(
            current_path=initial_path,
            on_selected=self._on_remote_folder_selected,
        )
        nav = self.get_ancestor(Adw.NavigationView)
        nav.push(page)

    def _on_remote_folder_selected(self, remote_path):
        self._remote_path = remote_path
        self.remote_folder_row.set_subtitle(remote_path)

    def _on_save_clicked(self, button):
        if not self._local_path or not self._remote_path:
            toast = Adw.Toast(title="Please select both local and remote folders")
            win = self.get_ancestor(Adw.ApplicationWindow)
            if hasattr(win, '_toast_overlay'):
                win._toast_overlay.add_toast(toast)
            return

        mode = 'mirror' if self.mirror_switch.get_active() else 'upload'
        db = get_db()

        if self._target:
            db.update_target(
                self._target['id'],
                local_path=self._local_path,
                remote_path=self._remote_path,
                mode=mode,
            )
        else:
            db.add_target(self._local_path, self._remote_path, mode)

        if self._targets_page:
            self._targets_page.refresh()

        nav = self.get_ancestor(Adw.NavigationView)
        nav.pop()

    def _on_delete_clicked(self, button):
        if not self._target:
            return

        dialog = Adw.AlertDialog(
            heading="Delete Target?",
            body=(
                f"This will remove the sync target for "
                f"{self._target['local_path']}. "
                f"Files on the server will not be deleted."
            ),
        )
        dialog.add_response("cancel", "Cancel")
        dialog.add_response("delete", "Delete")
        dialog.set_response_appearance(
            "delete", Adw.ResponseAppearance.DESTRUCTIVE
        )
        dialog.set_default_response("cancel")

        def on_response(dlg, response):
            if response == "delete":
                get_db().delete_target(self._target['id'])
                if self._targets_page:
                    self._targets_page.refresh()
                nav = self.get_ancestor(Adw.NavigationView)
                nav.pop()

        dialog.connect('response', on_response)
        dialog.present(self.get_ancestor(Gtk.Window))

File: src/target_edit_page.rs
use adw::prelude::*;
use adw::subclass::prelude::*;
use gtk::{gio, glib};
use std::cell::RefCell;

use crate::db::Target;
use crate::keyring;
use crate::remote_browser_page::SimplesyncRemoteBrowserPage;
use crate::window::SimplesyncWindow;

mod imp {
    use super::*;

    #[derive(Default, gtk::CompositeTemplate)]
    #[template(resource = "/io/github/nico359/simplesync/target_edit_page.ui")]
    pub struct SimplesyncTargetEditPage {
        #[template_child]
        pub local_row: TemplateChild<adw::ActionRow>,
        #[template_child]
        pub remote_row: TemplateChild<adw::ActionRow>,
        #[template_child]
        pub mirror_switch: TemplateChild<adw::SwitchRow>,
        #[template_child]
        pub save_button: TemplateChild<gtk::Button>,
        #[template_child]
        pub delete_button: TemplateChild<gtk::Button>,

        pub window: RefCell<Option<SimplesyncWindow>>,
        pub target_id: RefCell<Option<i64>>,
        pub local_path: RefCell<Option<String>>,
        pub remote_path: RefCell<Option<String>>,
        pub on_save: RefCell<Option<Box<dyn Fn() + 'static>>>,
    }

    impl std::fmt::Debug for SimplesyncTargetEditPage {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.debug_struct("SimplesyncTargetEditPage").finish()
        }
    }

    #[glib::object_subclass]
    impl ObjectSubclass for SimplesyncTargetEditPage {
        const NAME: &'static str = "SimplesyncTargetEditPage";
        type Type = super::SimplesyncTargetEditPage;
        type ParentType = adw::NavigationPage;

        fn class_init(klass: &mut Self::Class) {
            klass.bind_template();
        }

        fn instance_init(obj: &glib::subclass::InitializingObject<Self>) {
            obj.init_template();
        }
    }

    impl ObjectImpl for SimplesyncTargetEditPage {}
    impl WidgetImpl for SimplesyncTargetEditPage {}
    impl NavigationPageImpl for SimplesyncTargetEditPage {}
}

glib::wrapper! {
    pub struct SimplesyncTargetEditPage(ObjectSubclass<imp::SimplesyncTargetEditPage>)
        @extends gtk::Widget, adw::NavigationPage;
}

impl SimplesyncTargetEditPage {
    pub fn new() -> Self {
        glib::Object::builder().build()
    }

    pub fn set_window<F: Fn() + 'static>(
        &self,
        window: &SimplesyncWindow,
        target: Option<&Target>,
        on_save: F,
    ) {
        self.imp().window.replace(Some(window.clone()));
        self.imp().on_save.replace(Some(Box::new(on_save)));

        if let Some(target) = target {
            self.set_title("Edit Target");
            self.imp().target_id.replace(Some(target.id));
            self.imp().local_path.replace(Some(target.local_path.clone()));
            self.imp().remote_path.replace(Some(target.remote_path.clone()));
            self.imp().local_row.set_subtitle(&target.local_path);
            self.imp().remote_row.set_subtitle(&target.remote_path);
            self.imp().mirror_switch.set_active(target.mode == "mirror");
            self.imp().delete_button.set_visible(true);
        } else {
            self.set_title("Add Target");
        }

        self.setup_signals();
    }

    fn window(&self) -> SimplesyncWindow {
        self.imp().window.borrow().clone().expect("Window not set")
    }

    fn setup_signals(&self) {
        let page = self.clone();
        self.imp().local_row.connect_activated(move |_| {
            page.pick_local_folder();
        });

        let page = self.clone();
        self.imp().remote_row.connect_activated(move |_| {
            page.pick_remote_folder();
        });

        let page = self.clone();
        self.imp().save_button.connect_clicked(move |_| {
            page.save();
        });

        let page = self.clone();
        self.imp().delete_button.connect_clicked(move |_| {
            page.confirm_delete();
        });
    }

    fn pick_local_folder(&self) {
        let dialog = gtk::FileDialog::builder()
            .title("Select Local Folder")
            .build();

        let page = self.clone();
        dialog.select_folder(
            Some(&self.window()),
            None::<&gio::Cancellable>,
            move |result| {
                if let Ok(file) = result {
                    if let Some(path) = file.path() {
                        let path_str = path.to_string_lossy().to_string();
                        page.imp().local_row.set_subtitle(&path_str);
                        page.imp().local_path.replace(Some(path_str));
                    }
                }
            },
        );
    }

    fn pick_remote_folder(&self) {
        let creds = match keyring::load_credentials_sync() {
            Some(c) => c,
            None => {
                self.window().show_toast("No account configured");
                return;
            }
        };

        let browser = SimplesyncRemoteBrowserPage::new();
        let page = self.clone();
        browser.set_window(&self.window(), &creds, move |selected_path| {
            page.imp().remote_row.set_subtitle(&selected_path);
            page.imp().remote_path.replace(Some(selected_path));
        });
        self.window().navigation_view().push(&browser);
    }

    fn save(&self) {
        let local = self.imp().local_path.borrow().clone();
        let remote = self.imp().remote_path.borrow().clone();

        let (local, remote) = match (local, remote) {
            (Some(l), Some(r)) => (l, r),
            _ => {
                self.window().show_toast("Please select both local and remote folders");
                return;
            }
        };

        let mode = if self.imp().mirror_switch.is_active() { "mirror" } else { "upload" };

        let window = self.window();
        let db = window.db();
        let result = if let Some(id) = *self.imp().target_id.borrow() {
            db.update_target(id, &local, &remote, mode)
        } else {
            db.add_target(&local, &remote, mode).map(|_| ())
        };

        match result {
            Ok(()) => {
                if let Some(ref callback) = *self.imp().on_save.borrow() {
                    callback();
                }
                self.window().navigation_view().pop();
            }
            Err(e) => {
                self.window().show_toast(&format!("Error saving: {}", e));
            }
        }
    }

    fn confirm_delete(&self) {
        let target_id = match *self.imp().target_id.borrow() {
            Some(id) => id,
            None => return,
        };

        let dialog = adw::AlertDialog::builder()
            .heading("Delete Target?")
            .body("This will remove the sync target and its upload history.")
            .build();
        dialog.add_responses(&[("cancel", "Cancel"), ("delete", "Delete")]);
        dialog.set_response_appearance("delete", adw::ResponseAppearance::Destructive);
        dialog.set_default_response(Some("cancel"));

        let page = self.clone();
        dialog.connect_response(None, move |_, response| {
            if response == "delete" {
                let window = page.window();
                if let Err(e) = window.db().delete_target(target_id) {
                    page.window().show_toast(&format!("Error: {}", e));
                    return;
                }
                if let Some(ref callback) = *page.imp().on_save.borrow() {
                    callback();
                }
                page.window().navigation_view().pop();
            }
        });

        dialog.present(Some(&self.window()));
    }
}
